### DISCUSSION TITLE
Promover segurança com multi-contas e API KEY

### DISCUSSION THREAD

**Discussion by dansysanalyst** [tio-jobs/asaas-php#9] [Ideas]
Conforme a live https://www.youtube.com/watch?v=bifP6rNA4ow, fica a ideia de ter uma chave no .env que controle o uso automático da `access_key` estipulada no `.env`.

`ALLOW_SUB_ACCOUNTS=true/FALSE`

Se `ALLOW_SUB_ACCOUNTS` for `true` e o usuário não passar uma key vai estourar uma `exception` pedindo uma key.

Ou seja, se quando essa flag estiver ativada, não é permitido apenas o uso de chave do .env. e o usuário sempre terá que informar uma key um método fluente como `->withKey($apiKey)`

**Comment by dansysanalyst** [comment]
No meu ponto de vista, o nome `ALLOW_SUB_ACCOUNTS` não condiz totalmente com o comportamento.


**ALLOW_SUB_ACCOUNTS=false**

`asaasPhp()->foo()` 
Resultado: OK


**ALLOW_SUB_ACCOUNTS=true**

`asaasPhp()->foo()` 
Resultado: exception "the API key is required".


**ALLOW_SUB_ACCOUNTS=true**

`asaasPhp()->withKey($apiKey)->foo()` 
Resultado: OK (temos uma key que satisfaz a configuração)


**ALLOW_SUB_ACCOUNTS=false**

`asaasPhp()->withKey($apiKey)->foo()` 

Nesse caso, eu passo uma key, mas a configuração não permite uma sub-conta. 

Chegamos a um impasse. Ele vai ler a chave do .env?  Vai usar a do withKey()? ou vai estourar exception?

Se temos que sub-contas não são permitidas, ele não devia poder usar esse método.

<hr>

Sugiro um nome como `ALWAYS_REQUIRE_API_KEY=true/FALSE`.  

A documentação pode dizer:  _Para sua comodidade, o pacote usa por padrão a chave do .env. Para trabalhar com multi-contas sugerimos configurar ALWAYS_REQUIRE_API_KEY=true_.  A parte de sub-contas na documentação também pode uma mensagem similar.

**Reply by icarojobs** [reply]
Creio que trabalhar com a chave `ALWAYS_REQUIRE_API_KEY=true | false` também faz mais sentido. Só precisamos detalhar bem o que exatamente essa chave faz, e quais são os benefícios para projetos que trabalham com multi-contas.

Show de bola! Gostei muito disso!

**Comment by icarojobs** [comment]
> Conforme a live https://www.youtube.com/watch?v=bifP6rNA4ow, fica a ideia de ter uma chave no .env que controle o uso automático da `access_key` estipulada no `.env`.
> 
> `ALLOW_SUB_ACCOUNTS=true/FALSE`
> 
> Se `ALLOW_SUB_ACCOUNTS` for `true` e o usuário não passar uma key vai estourar uma `exception` pedindo uma key.
> 
> Ou seja, se quando essa flag estiver ativada, não é permitido apenas o uso de chave do .env. e o usuário sempre terá que informar uma key um método fluente como `->withKey($apiKey)`

Boa Dan! obrigado por ajudar a ativar o discussions do projeto e já trazer esse ponto, que é extremamente importante, antes do lançamento oficial do pacote!
